onnxruntime/core/providers/coreml/builders/impl/pad_op_builder.cc

```diff
@@ -0,0 +1,170 @@
+#endif
+
+  // Operator support related
+ private:
+  bool IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                         const logging::Logger& logger) const override;
+
+  int GetMinSupportedOpSet(const Node& /* node */) const override {
+    // Note: before Pad-11, inputs `pads` and `constant_value` were attributes
+    return 11;
+  }
+};
+
+// Add operator related
+
+#ifdef __APPLE__
+void PadOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  model_builder.AddInitializerToSkip(node.InputDefs()[1]->Name());  //  pads
+  model_builder.AddInitializerToSkip(node.InputDefs()[2]->Name());  //  constant_value
+}
+
+Status PadOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                           const Node& node,
+                                           const logging::Logger& logger) const {
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(model_builder, node);
+
+  auto* coreml_pad = layer->mutable_padding();
+  auto* constant_padding_type = coreml_pad->mutable_constant();  // CoreML::Specification::PaddingLayerParams_PaddingConstant
+
+  const auto& input_defs = node.InputDefs();
+
+  const auto& pads_tensor = *model_builder.GetInitializerTensors().at(input_defs[1]->Name());            // pads
+  const auto& constant_value_tensor = *model_builder.GetInitializerTensors().at(input_defs[2]->Name());  // constant_value
+
+  float pad_value = 0.0f;
+  Initializer pad_value_raw_data_init(constant_value_tensor);
+  pad_value = pad_value_raw_data_init.DataAsSpan<float>()[0];
+  constant_padding_type->set_value(pad_value);
+
+  Initializer pads_initializer_raw_data(pads_tensor);
+  auto pads_span = pads_initializer_raw_data.DataAsSpan<int64_t>();
+
+  // Add padding
+  auto* height_border = coreml_pad->mutable_paddingamounts()->add_borderamounts();
+  height_border->set_startedgesize(pads_span[0]);
+  height_border->set_endedgesize(pads_span[2]);
+  auto* width_border = coreml_pad->mutable_paddingamounts()->add_borderamounts();
+  width_border->set_startedgesize(pads_span[1]);
+  width_border->set_endedgesize(pads_span[3]);
+
+  *layer->mutable_input()->Add() = input_defs[0]->Name();
+  *layer->mutable_output()->Add() = node.OutputDefs()[0]->Name();
+
+  model_builder.AddLayer(std::move(layer));
+
+  return Status::OK();
+}
+#endif
+
+// Operator support related
+
+bool PadOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                                     const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  const auto& initializers = input_params.graph_viewer.GetAllInitializedTensors();
+
+  {
+    std::vector<int64_t> input_shape;
+    if (!GetShape(*input_defs[0], input_shape, logger))
+      return false;
+
+    if (input_shape.size() > 4 || input_shape.empty()) {
```

Comment:
from the [spec](https://apple.github.io/coremltools/mlmodel/Format/NeuralNetwork.html#paddinglayerparams), regarding the input:

"For ranks greater than 2, the leading dimensions, starting from 0 to -4 (inclusive), are all treated as batch. That is, padding is applied on last two dimensions."

so we should only support padding the last 2D right? (also, I'm not sure where they are getting "-4" from.)

the check below for `pads` input to have length 4 may not be sufficient, e.g., if an `axes` input specifies dimensions other than the last two.

Reply:
make sense. thanks for suggesting this.

and I guess `0 to -4 (inclusive) leading dimensions` is just a similar general condition as the other CoreML operators - for pad inputs have ranks even greater or equal than 4, all leading ones are treated as batch.
 
PaddinglayerParams only requires rank greater than 2, and I agree that the padding should be only applied on the last 2D - [H,W].

but should we limit the input_shape.size() == 2? Or can we allow supporting input_shape.size() between 2-4d, and limit the support for specifying the optional `axes` input? As CoreML does support input cases greater than 2D, and uses BorderAmounts type for adding the pads. https://apple.github.io/coremltools/mlmodel/Format/NeuralNetwork.html#borderamounts.
and it works with other input shape > 2 case. (for example conv.)

Reply:
update: added additional checking to ensure we only support non-zero values in `pads`tensor on the last two dimensions. i.e. at index input_rank - 1, input_rank - 2 and 2*(input_rank - 1), 2*(input_rank - 2).

Reply:
we should also check the `axes` input to verify that at least the last two dimensions are specified.

Reply:
I think the coreml spec is wrong and means `0 to -3 (inclusive)` given only the last 2 dims are padded.